aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -0,0 +1,209 @@
+    // TH and ATen. We should remove the TH gesv bindings, especially
+    // since the lapackGesv function is already in ATen.
+    return at::_gesv_single(self, A);
+  }
+
+  gesvCheckInputs(self, A);
+
+  // broadcast the batch dimensions of self and A.
+  IntList self_batch_sizes(self.sizes().data(), self.ndimension() - 2);
+  IntList A_batch_sizes(A.sizes().data(), A.ndimension() - 2);
+  std::vector<int64_t> expand_batch_portion = infer_size(self_batch_sizes, A_batch_sizes);
+
+  std::vector<int64_t> self_expand_size({expand_batch_portion});
+  self_expand_size.insert(self_expand_size.end(), { self.size(-2), self.size(-1) });
+
+  std::vector<int64_t> A_expand_size({expand_batch_portion});
+  A_expand_size.insert(A_expand_size.end(), { A.size(-2), A.size(-1) });
+
+  Tensor self_broadcasted  = self.expand(self_expand_size);
+  Tensor A_broadcasted = A.expand(A_expand_size);
+  return at::_gesv_helper(self_broadcasted, A_broadcasted);
+}
+
+std::tuple<Tensor&,Tensor&> gesv_out(Tensor& solution, Tensor& lu, const Tensor& self, const Tensor& A) {
+  AT_CHECK(self.dim() == 2 && A.dim() == 2, 
+           "torch.gesv() with the `out` keyword does not support batching. "
+           "b.dim() (", self.dim(), ") and A.dim() (", A.dim(), ") must both be 2.");
+  return at::_gesv_single_out(solution, lu, self, A);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ inverse ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_inverse(Tensor& self, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("inverse: LAPACK library not found in compilation");
+#endif
+  auto self_data = self.data<scalar_t>();
+  auto self_matrix_stride = matrixStride(self);
+
+  auto batch_size = batchCount(self);
+  auto n = self.size(-2);
+
+  auto ipiv = at::empty({n}, self.type().toScalarType(kInt));
+  int lwork;
+  scalar_t wkopt;
+  Tensor work;
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    int info;
+    scalar_t* self_working_ptr = &self_data[i * self_matrix_stride];
+    lapackGetrf<scalar_t>(n, n, self_working_ptr, n, ipiv.data<int>(), &info);
+    infos[i] = info;
+    if (info != 0) {
+      return;
+    }
+
+    // Run twice, first to get the optimum work size
+    lwork = -1;
+    lapackGetri<scalar_t>(n, self_working_ptr, n, ipiv.data<int>(), &wkopt, lwork, &info);
+
+    lwork = static_cast<int>(wkopt);
+    work = at::empty({lwork}, self.type());
+
+    // now to compute the actual inverse
+    lapackGetri<scalar_t>(n, self_working_ptr, n, ipiv.data<int>(), work.data<scalar_t>(), lwork, &info);
+    infos[i] = info;
+    if (info != 0) {
+      return;
+    }
+  }
+}
+
+LINALG_HELPER_1_ARGS(inverse, self, cpu)
+
+Tensor inverse(const Tensor &self) {
+  if (self.size(-1) == 0) {
+    return at::empty_like(self);
+  }
+  if (self.dim() == 2) {
+    return at::_getri_single(self);
+  }
+  inverseCheckInputs(self);
+  return at::_inverse_helper(self);
```

Comment:
Hmm would prefer this be more explicit that it was generated from LINALG_HELPER_1_ARGS.
Could we name the function `::_generate_inverse_linalg_helper` and the macro `GENERATE_LINALG_HELPER_1_ARGS` or something like that?

Reply:
Oh, I see. _inverse_helper isn't generated; _inverse_helper_cuda / inverse_helper_cpu are. Could we rename those so that it is clearer that they were generated?

Reply:
Thanks. I'll rename the macros to `GENERATE_LINALG_HELPER_n_ARGS`.